src/optimization/DAG.cpp

```diff
@@ -0,0 +1,149 @@
+//
+// Created by nomaddo on 17/12/18.
+//
+
+#include <log.h>
+#include <assert.h>
+#include <iostream>
+#include <functional>
+#include <algorithm>
+#include <string>
+#include <stack>
+#include <utility>
+#include <vector>
+
+#include "DAG.h"
+#include "../InstructionWalker.h"
+#include "log.h"
+
+#include "../ControlFlowGraph.h"
+#include "../DebugGraph.h"
+#include "../InstructionWalker.h"
+#include "../Profiler.h"
+
+namespace vc4c {
+
+/* XXX should be replaced with common routines that generate such instructions? */
+/* reading from mutex I/O register means getting mutex */
+static bool isGetLocInstruction (IL * instr) {
+	for (Value v: instr->getArguments()) {
+		if (v.hasRegister(REG_MUTEX)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/* writing to mutex I/O register means release of mutex */
+static bool isReleaseLocInstruction (IL * instr) {
```

Comment:
Same here: `instr->writesRegister(REG_MUTEX)`.

NOTE: Both methods do not regard combined-instructions, but mutex-locks and mutex-releases are never combined anyway.